src/psij/executors/batch/batch_scheduler_executor.py

```diff
@@ -639,45 +640,51 @@ def __init__(self, name: str, config: BatchSchedulerExecutorConfig,
         self.name = name
         self.daemon = True
         self.config = config
-        self.executor = executor
+        self.done = False
+        self.executor = weakref.ref(executor, self._stop)
         # native_id -> job
         self._jobs: Dict[str, List[Job]] = {}
         # counts consecutive errors while invoking qstat or equivalent
         self._poll_error_count = 0
         self._jobs_lock = RLock()
 
     def run(self) -> None:
-        logger.debug('Executor %s: queue poll thread started', self.executor)
+        logger.debug('Executor %s: queue poll thread started', self.executor())
         time.sleep(self.config.initial_queue_polling_delay)
-        while True:
+        while not self.done:
             self._poll()
             time.sleep(self.config.queue_polling_interval)
 
+    def _stop(self, exec: object) -> None:
+        self.done = True
+
     def _poll(self) -> None:
+        exec = self.executor()
+        if exec is None:
+            return
         with self._jobs_lock:
             if len(self._jobs) == 0:
                 return
             jobs_copy = dict(self._jobs)
         logger.info('Polling for %s jobs', len(jobs_copy))
         try:
-            out = self.executor._run_command(self.executor.get_status_command(jobs_copy.keys()))
+            if exec:
```

Comment:
`if exec` is not needed, that is checked above.

Reply:
Ah, thank you. Will issue a small PR for that.